lms/static/js/spec/search/search_spec.js

```diff
@@ -0,0 +1,489 @@
+            this.listView.render();
+            this.listView.loadNext();
+            expect(this.listView.$el.find('a.search-load-next .icon')[0]).toBeVisible();
+            this.listView.renderNext();
+            expect(this.listView.$el.find('a.search-load-next .icon')[0]).toBeHidden();
+        });
+
+    });
+
+
+    describe('SearchRouter', function () {
+
+        beforeEach(function () {
+            this.router = new SearchRouter();
+        });
+
+        it ('has a search route', function () {
+            expect(this.router.routes['search/:query']).toEqual('search');
+        });
+
+    });
+
+
+    describe('SearchApp', function () {
+
+        beforeEach(function () {
+            TemplateHelpers.installTemplates([
+                'templates/courseware_search/search_item',
+                'templates/courseware_search/search_list',
+                'templates/courseware_search/search_loading',
+                'templates/courseware_search/search_error'
+            ], true);
+
+            // spy on these methods before they are bound to events
+            spyOn(SearchRouter.prototype, 'navigate').andCallThrough();
+            spyOn(SearchForm.prototype, 'doSearch').andCallThrough();
+            spyOn(SearchCollection.prototype, 'performSearch').andCallThrough();
+            spyOn(SearchCollection.prototype, 'cancelSearch').andCallThrough();
+            spyOn(SearchCollection.prototype, 'loadNextPage').andCallThrough();
+            spyOn(SearchListView.prototype, 'showLoadingMessage').andCallThrough();
+            spyOn(SearchListView.prototype, 'clear').andCallThrough();
+
+            this.app = new SearchApp('a/b/c');
+
+        });
+
+        it ('has all necessary components', function () {
+            expect(this.app).toBeDefined();
+            expect(this.app.router).toBeDefined();
+            expect(this.app.form).toBeDefined();
+            expect(this.app.collection).toBeDefined();
+            expect(this.app.results).toBeDefined();
+        });
+
+        it ('shows loading message on search', function () {
+            this.app.form.trigger('search');
+            expect(this.app.results.showLoadingMessage).toHaveBeenCalled();
+        });
+
+        it ('performs search', function () {
+            this.app.form.trigger('search');
+            expect(this.app.collection.performSearch).toHaveBeenCalled();
+        });
+
+        it ('updates navigation history on search', function () {
+            this.app.form.trigger('search', 'search_string');
+            expect(this.app.router.navigate).toHaveBeenCalledWith('search/search_string', { replace: true });
+        });
+
+        it ('cancels search', function () {
+            this.app.form.trigger('clear');
+            expect(this.app.collection.cancelSearch).toHaveBeenCalled();
+        });
+
+        it ('clears results', function () {
+            this.app.form.trigger('clear');
+            expect(this.app.results.clear).toHaveBeenCalled();
```

Comment:
I don't like all these tests that verify that an internal function has been called. They don't really seem like unit tests to me. I'd rather that you not use all the spies, but instead verify that the actual method produces the desired results.